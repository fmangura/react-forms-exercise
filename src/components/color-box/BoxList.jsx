import React, { useState } from "react";
import NewBoxForm from './NewBoxForm'
import Box from './Box'
import { v4 as uuidv4 } from 'uuid';


function BoxList() {
    const [boxes, setBoxes] = useState([])
    
    const addBox = (newbox) => {
        setBoxes(boxes => [...boxes, {...newbox, id:uuidv4() }])
    }

    const rmvBox = (id) => {
        setBoxes(boxes => {
            const copyBox = [...boxes];
            return copyBox.filter(box => box.id !== id)
        })
    }

    return (
        <div>
            <div>
              <NewBoxForm addBox={addBox}/>  
            </div>
            <h1>Boxes:</h1>
            <div className="Boxes">
                {boxes.map(({color, width, height, id}) => <Box key={id} color={color} width={width} height={height} rmvBox={rmvBox} id={id} />)}
            </div>
        </div>
    )
}

export default BoxList