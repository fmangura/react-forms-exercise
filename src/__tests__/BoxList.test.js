import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom'
import BoxList from '../components/color-box/BoxList';

it('renders properly', function() {
    render(<BoxList />);
})

it('Matches snapshot', function() {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
})

it('Correct info are rendering', function() {
    const { getByText } = render(<BoxList />);
    
    expect(getByText('Boxes:', { exact: false })).toBeInTheDocument();
})

it('can get new boxes', function() {
    const {getByText, getByLabelText} = render(<BoxList />);

    //Checks a box does not exist
    expect(document.querySelector('.aBox')).not.toBeInTheDocument();

    const colorInput = getByLabelText("Color:", { exact: false });
    const widthInput = getByLabelText("Width px:", { exact: false });
    const HeightInput = getByLabelText("Height px:", { exact: false });
    const submit = getByText('Add Box');

    fireEvent.change(colorInput, { target: { value: "blue" }});
    fireEvent.change(widthInput, { target: { value: 50 }});
    fireEvent.change(HeightInput, { target: { value: 50 }});
    fireEvent.click(submit);

    //Checks that a box exists after the form is submitted
    expect(document.querySelector('.aBox')).toBeInTheDocument();

})

it('can delete box', function() {
    const {getByText, getByLabelText} = render(<BoxList />);

    const colorInput = getByLabelText("Color:", { exact: false });
    const widthInput = getByLabelText("Width px:", { exact: false });
    const HeightInput = getByLabelText("Height px:", { exact: false });
    const submit = getByText('Add Box');

    fireEvent.change(colorInput, { target: { value: "blue" }});
    fireEvent.change(widthInput, { target: { value: 50 }});
    fireEvent.change(HeightInput, { target: { value: 50 }});
    fireEvent.click(submit);

    //Checks that the box was made
    const box = document.querySelector('.aBox')
    expect(box).toBeInTheDocument();
    
    // Clicks box
    fireEvent.click(box);

    //Checks box was deleted
    expect(box).not.toBeInTheDocument();
})