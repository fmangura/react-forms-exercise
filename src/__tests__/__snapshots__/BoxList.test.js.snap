// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="color"
        >
          Color: 
        </label>
        <input
          id="color"
          name="color"
          placeholder="Color"
          type="text"
          value=" "
        />
        <label
          for="width"
        >
          Width px: 
        </label>
        <input
          id="width"
          name="width"
          placeholder="Width px"
          type="integer"
          value=" "
        />
        <label
          for="height"
        >
          Height px: 
        </label>
        <input
          id="height"
          name="height"
          placeholder="Height px"
          type="integer"
          value=" "
        />
        <button>
          Add Box
        </button>
      </form>
    </div>
    <h1>
      Boxes:
    </h1>
    <div
      class="Boxes"
    />
  </div>
</DocumentFragment>
`;
